paper-api/src/main/java/io/papermc/paper/registry/data/dialog/DialogBase.java

```diff
@@ -0,0 +1,208 @@
+    static DialogBase create(
+        final Component title,
+        final @Nullable Component externalTitle,
+        final boolean canCloseWithEscape,
+        final boolean pause,
+        final DialogAfterAction afterAction,
+        final List<? extends DialogBody> body,
+        final List<DialogInput> inputs
+    ) {
+        return new DialogBaseImpl(title, externalTitle, canCloseWithEscape, pause, afterAction, List.copyOf(body), inputs);
+    }
+
+    /**
+     * Creates a new dialog base builder.
+     *
+     * @param title the title of the dialog
+     * @return a new dialog base builder
+     */
+    @Contract(value = "_ -> new", pure = true)
+    static Builder builder(final Component title) {
+        return new DialogBaseImpl.BuilderImpl(title);
+    }
+
+    /**
+     * The title of the dialog.
+     *
+     * @return the title
+     */
+    @Contract(pure = true)
+    Component title();
+
+    /**
+     * The external title of the dialog. This title
+     * is used on buttons that open this dialog.
+     *
+     * @return the external title or null
+     */
+    @Contract(pure = true)
+    @Nullable Component externalTitle();
+
+    /**
+     * Returns if this dialog can be closed with the "escape" keybind.
+     *
+     * @return if the dialog can be closed with "escape"
+     */
+    @Contract(pure = true)
+    boolean canCloseWithEscape();
+
+    /**
+     * Returns if this dialog should pause the game when opened (single-player only).
+     *
+     * @return if the dialog pauses the game
+     */
+    @Contract(pure = true)
+    boolean pause();
+
+    /**
+     * The action to take after the dialog is closed.
+     *
+     * @return the action to take after the dialog is closed
+     */
+    @Contract(pure = true)
+    DialogAfterAction afterAction();
+
+    /**
+     * The body of the dialog.
+     * <p>
+     * The body is a list of {@link DialogBody} elements that will be displayed in the dialog.
+     *
+     * @return the body of the dialog
+     */
+    @Contract(pure = true)
+    @Unmodifiable List<DialogBody> body();
+
+    /**
+     * The inputs of the dialog.
+     * <p>
+     * The inputs are a list of {@link DialogInput} elements that will be displayed in the dialog.
+     *
+     * @return the inputs of the dialog
+     */
+    @Contract(pure = true)
+    @Unmodifiable List<DialogInput> inputs();
+
+    /**
+     * Actions to take after the dialog is closed.
+     */
+    enum DialogAfterAction {
+        /**
+         * Closes the dialog and returns to the previous non-dialog screen (if any).
+         */
+        CLOSE("close"),
+        /**
+         * Does nothing (keeps the current screen open).
+         */
+        NONE("none"),
+        /**
+         * Replaces dialog with a "waiting for response" screen.
+         */
+        WAIT_FOR_RESPONSE("wait_for_response");
+
+        public static final Index<String, DialogAfterAction> NAMES = Index.create(DialogAfterAction.class, e -> e.name);
```

Comment:
The public static Index NAMES on enums can be confusing depending on how high up the IDE puts it, I'd maybe make it private with a getter instead, but I have no strong opinion on that